e2e-test-server/server.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2021 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import os
+from typing import Iterator
+
+from flask import Flask, Response, request
+from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.sdk.trace.sampling import ALWAYS_ON
+from opentelemetry.trace import Tracer
+
+TEST_ID = "test-id"
+INSTRUMENTING_MODULE_NAME = "opentelemetry-ops-e2e-test-server"
+
+app = Flask(__name__)
+
+
+@contextlib.contextmanager
+def common_setup() -> Iterator[tuple[str, Tracer]]:
+    """\
+    Context manager with common setup for test endpoints
+
+    It extracts the test-id header, creates a tracer, and finally flushes
+    spans created during the test
+    """
+
+    if TEST_ID not in request.headers:
```

Comment:
Does this imply we don't set up the tracer provider until the first URL comes in OR that we set one up for EVERY request to the server?

Reply:
We set one up for every request to the server. So test has its own provider/processor/exporter for isolation

Reply:
Interesting.  Is that a requirement for other languages?

E.g. would it be ok to just "flush" after serving the trace?

Reply:
It is in the design, but flush would work too. I figured each test scenario might as well have a completely separate SDK setup.

Do you have a preference for flush?

Reply:
No, just worried about other languages.  What's here is good for now.

Reply:
Talked offline – each BatchSpanProcessor creates a new thread. But, calling `tracer_provider.shutdown()` will cause the thread to end and wait block until `Thread.join()` finishes.

Updated the spec that this is not a requirement but a suggestion.